File: src/Types/Block.hs
{-# LANGUAGE DeriveGeneric #-}

module Types.Block where

import GHC.Generics (Generic)
import Data.Aeson (ToJSON, FromJSON)
import Hashing.Hash (hashString)
import Hashing.Serialization (serializePreBlock)
import Types.PreBlock

data Block = Block
  { blockContent :: PreBlock
  , hashValue :: String
  } deriving (Show, Eq, Generic)

instance ToJSON Block
instance FromJSON Block

createBlock :: PreBlock -> Block
createBlock pb =
  Block
    { blockContent = pb
    , hashValue = hashString (serializePreBlock pb)
    }

File: bin/main/block/Block.kt
package block

data class Block(
    private val preBlock: PreBlock,
    private val hash: String,
) {
    fun getIndex(): Int = preBlock.getIndex()
    fun getTimestamp(): Long = preBlock.getTimestamp()
    fun getTransactions() = preBlock.getTransactions()
    fun getPreviousHash(): String = preBlock.getPreviousHash()
    fun getNonce(): Long = preBlock.getNonce()
    fun getHash(): String = hash
}
